src/flag_gems/ops/softmax.py

```diff
@@ -10,106 +10,286 @@
-        triton.Config({"BLOCK_M": 2}, num_stages=5),
-        triton.Config({"BLOCK_M": 4}, num_stages=4),
-        triton.Config({"BLOCK_M": 4}, num_stages=5),
-        triton.Config({"BLOCK_M": 8}, num_stages=4),
-        triton.Config({"BLOCK_M": 8}, num_stages=5),
+        triton.Config({"TILE_K": 32}),
+        triton.Config({"TILE_K": 64}),
+        triton.Config({"TILE_K": 128}),
+        triton.Config({"TILE_K": 256}),
+        triton.Config({"TILE_K": 512}),
+        triton.Config({"TILE_K": 1024}),
     ],
     key=[
-        "M",
-        "N",
+        "K",
     ],
 )
 @triton.heuristics(
     values={
-        "BLOCK_N": lambda args: triton.next_power_of_2(args["N"]),
-        "num_warps": lambda args: (
-            4 if args["N"] <= 1024 else (8 if args["N"] <= 2048 else 16)
-        ),
+        "TILE_N": lambda args: max(1, 1024 // args["TILE_K"]),
+        "ONE_TILE_PER_CTA": lambda args: args["TILE_N"] >= args["N"],
     },
 )
 @triton.jit
-def softmax_kernel(
+def softmax_kernel_non_inner(
     output_ptr,
     input_ptr,
     M,
     N,
     K,
-    BLOCK_M: tl.constexpr,
-    BLOCK_N: tl.constexpr,
+    TILE_N: tl.constexpr,
+    TILE_K: tl.constexpr,
+    ONE_TILE_PER_CTA: tl.constexpr,
 ):
-    pid_m = tl.program_id(0)
     pid_k = tl.program_id(1)
-    m_offset = pid_m * BLOCK_M + tl.arange(0, BLOCK_M)
-    n_offset = tl.arange(0, BLOCK_N)
-    offset = m_offset[:, None] * N * K + n_offset[None, :] * K + pid_k
-    mask = m_offset[:, None] < M and n_offset[None, :] < N
-    input_ptrs = input_ptr + offset
-    inp = tl.load(input_ptrs, mask=mask, other=-float("inf"))
-    row_minus_max = inp - tl.max(inp, axis=1)[:, None]
-    numerator = tl.exp(row_minus_max)
-    denominator = tl.sum(numerator, axis=1)[:, None]
-    softmax_output = numerator / denominator
-    output_ptrs = output_ptr + offset
-    tl.store(output_ptrs, softmax_output, mask=mask)
+    pid_m = tl.program_id(0)
+
+    k_offsets = pid_k * TILE_K + tl.arange(0, TILE_K)
+    if ONE_TILE_PER_CTA:
+        n_offsets = tl.arange(0, TILE_N)
+        offset = pid_m * N * K + n_offsets[:, None] * K + k_offsets
+        mask = (n_offsets[:, None] < N) & (k_offsets < K)
+        input_ptrs = input_ptr + offset
+        inp = tl.load(input_ptrs, mask=mask, other=-float("inf"))
+        m = tl.max(inp, 0)
+        e = tl.exp(inp - m[None, :])
+        z = tl.sum(e, 0)
+        out = e / z
+        output_ptrs = output_ptr + offset
+        tl.store(output_ptrs, out, mask=mask)
+    else:
+        m = tl.full([TILE_K], value=float("-inf"), dtype=tl.float32)
+        z = tl.full([TILE_K], value=0.0, dtype=tl.float32)
+
+        n_offsets = tl.arange(0, TILE_N)
+        offset = pid_m * N * K + n_offsets[:, None] * K + k_offsets
+        for _ in range(0, N, TILE_N):
+            mask = (n_offsets[:, None] < N) & (k_offsets < K)
+            input_ptrs = input_ptr + offset
+            inp = tl.load(input_ptrs, mask=mask, other=-float("inf"))
```

Comment:
This should be a slight improvement in performance.
```suggestion
            inp = tl.load(input_ptrs, mask=mask, other=-float("inf"), eviction_policy="evict_last")
```

Reply:
Did you see any improvement in real tests? I'm just curious. All blocks are loaded with eviction last hints is no different than LRU I guess.

Reply:
Here are my test results on the 4090 with shape=(64, N), dtype=float16

![image](https://github.com/FlagOpen/FlagGems/assets/58718936/69cc5860-0a6f-4233-9919-e9fea42de2b6)

Reply:
Does it take into account the evict_first loads also? They do improve perf.

Reply:
use `evict_first` in the first loop will lead to a lower performance

Reply:
No I wasn't referring to the first loop. The evict_first policy in the second loop seems to have more clear impact.